File: jar-client/src/main/java/com/beancounter/client/services/TrnService.java
package com.beancounter.client.services;

import com.beancounter.auth.common.TokenService;
import com.beancounter.common.contracts.TrnRequest;
import com.beancounter.common.contracts.TrnResponse;
import com.beancounter.common.input.TrustedTrnQuery;
import com.beancounter.common.model.Portfolio;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestHeader;

@Service
public class TrnService {
  private final TrnGateway trnGateway;
  private final TokenService tokenService;

  TrnService(TrnGateway trnGateway, TokenService tokenService) {
    this.trnGateway = trnGateway;
    this.tokenService = tokenService;
  }

  public TrnResponse write(TrnRequest trnRequest) {
    return trnGateway.write(tokenService.getBearerToken(), trnRequest);
  }

  // Figure out service to service tokens
  public TrnResponse query(TrustedTrnQuery trustedTrnQuery) {
    return trnGateway.read(tokenService.getBearerToken(), trustedTrnQuery);
  }

  public TrnResponse query(Portfolio portfolio) {
    return trnGateway.read(tokenService.getBearerToken(), portfolio.getId());
  }

  @FeignClient(name = "trns",
      url = "${marketdata.url:http://localhost:9510/api}")
  public interface TrnGateway {
    @PostMapping(value = "/trns",
        produces = {MediaType.APPLICATION_JSON_VALUE},
        consumes = {MediaType.APPLICATION_JSON_VALUE})
    TrnResponse write(
        @RequestHeader("Authorization") String bearerToken,
        TrnRequest trnRequest);

    @GetMapping(value = "/trns/portfolio/{portfolioId}",
        produces = {MediaType.APPLICATION_JSON_VALUE})
    TrnResponse read(
        @RequestHeader("Authorization") String bearerToken,
        @PathVariable("portfolioId") String portfolioId);


    @PostMapping(value = "/trns/query",
        produces = {MediaType.APPLICATION_JSON_VALUE},
        consumes = {MediaType.APPLICATION_JSON_VALUE})
    TrnResponse read(
        @RequestHeader("Authorization") String bearerToken,
        TrustedTrnQuery trnQuery);

  }

}

File: jar-client/src/main/kotlin/com/beancounter/client/services/TrnService.kt
package com.beancounter.client.services

import com.beancounter.auth.TokenService
import com.beancounter.common.contracts.TrnRequest
import com.beancounter.common.contracts.TrnResponse
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.input.TrustedTrnQuery
import com.beancounter.common.model.Portfolio
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClient

/**
 * Obtain transactions from the backend.
 */
@Service
class TrnService(
    @Qualifier("bcDataRestClient")
    private val restClient: RestClient,
    private val tokenService: TokenService
) {
    fun write(trnRequest: TrnRequest): TrnResponse =
        restClient
            .post()
            .uri("/api/trns")
            .header(HttpHeaders.AUTHORIZATION, tokenService.bearerToken)
            .contentType(MediaType.APPLICATION_JSON)
            .body(trnRequest)
            .retrieve()
            .body(TrnResponse::class.java)
            ?: throw BusinessException("Failed to write transactions")

    @CircuitBreaker(name = "default")
    fun query(trustedTrnQuery: TrustedTrnQuery): TrnResponse =
        restClient
            .post()
            .uri("/api/trns/query")
            .header(HttpHeaders.AUTHORIZATION, tokenService.bearerToken)
            .contentType(MediaType.APPLICATION_JSON)
            .body(trustedTrnQuery)
            .retrieve()
            .body(TrnResponse::class.java)
            ?: throw BusinessException("Failed to query transactions")

    fun query(
        portfolio: Portfolio,
        asAt: String = "today"
    ): TrnResponse =
        restClient
            .get()
            .uri("/api/trns/portfolio/{portfolioId}/{asAt}", portfolio.id, asAt)
            .header(HttpHeaders.AUTHORIZATION, tokenService.bearerToken)
            .retrieve()
            .body(TrnResponse::class.java)
            ?: throw BusinessException("Failed to query portfolio transactions")
}